### PR TITLE
Refactor add_<declare/invoke/deploy_account>_transactions

### LINKED ISSUE #349: Transaction with query version executes successfully Instead of throwing error
**Describe the bug (observed vs expected behavior)**

Currently, the devnet allows the execution of a transaction that is meant to be used for fee estimation instead of throwing an error.

**Not reproducible on alpha-goerli**

- [X] This issue is only present on Devnet and cannot be reproduced on alpha-goerli (check the box if true).

**To Reproduce**

1. Create any `Invoke`/`Declare`/`DeployAccount` transaction.
2. Sign the transaction off-chain for fee estimation. Use the query version (`version` + `2**128`), for example `0x100000000000000000000000000000001`.
3. Proceed to execute the transaction as if it's a normal transaction.
4. At this point, the process should produce an error. However, it successfully completes the execution instead.

**Devnet version**

- I am using Devnet version: [ebd9f85](https://github.com/0xSpaceShard/starknet-devnet-rs/tree/ebd9f8549302f91a9071fbecf69e31142554ac14)

**System specifications**

- OS: MacOS M1

### PR DESCRIPTION
**Pull request by marioiordanov**
## Usage related changes

closes #349 

## Development related changes

Refactor add_transaction methods. By combining handling cases of transaction versions
Replaced thread::sleep with Tokio::time::sleep in async methods
Added check for returning error when a query_only transaction is send in write endpoints

## Checklist:

- [ ] Checked the relevant parts of [development docs](https://github.com/0xSpaceShard/starknet-devnet-rs/?tab=readme-ov-file#development)
- [ ] Applied formatting - `./scripts/format.sh`
- [ ] No linter errors - `./scripts/clippy_check.sh`
- [ ] No unused dependencies - `./scripts/check_unused_deps.sh`
- [ ] Performed code self-review
- [ ] Rebased to the latest commit of the target branch (or merged it into my branch)
- [ ] Updated the docs if needed, including the [TODO section](https://github.com/0xSpaceShard/starknet-devnet-rs/?tab=readme-ov-file#todo-to-reach-feature-parity-with-the-pythonic-devnet)
- [ ] Linked the issues which this PR resolves
- [ ] Updated the tests if needed; all passing ([execution info](https://github.com/0xSpaceShard/starknet-devnet-rs/?tab=readme-ov-file#test-execution))

### CODE REVIEW

File: crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v1.rs [hunk 1]
```diff
@@ -72,11 +73,19 @@ impl BroadcastedDeclareTransactionV1 {
         let class_info: ClassInfo =
             ContractClass::Cairo0(self.contract_class.clone()).try_into()?;
 
-        Ok(DeclareTransaction::new(
-            sn_api_declare,
-            starknet_api::transaction::TransactionHash(transaction_hash.into()),
-            class_info,
-        )?)
+        if FieldElement::from(self.common.version) > QUERY_VERSION_OFFSET {
```
Comment (FabijanC) [inline crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v1.rs:76]: Should it be `>` or `>=`? Also applicable to other occurrences of comparison with `QUERY_VERSION_OFFSET`. In other words, what about transactions with version 0 (i.e. their query counterparts)?

File: crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v2.rs [hunk 2]
```diff
@@ -76,11 +77,19 @@ impl BroadcastedDeclareTransactionV2 {
...
+                starknet_api::transaction::TransactionHash(txn_hash.into()),
+                ContractClass::Cairo1(self.contract_class.clone()).try_into()?,
+            )?)
+        } else {
+            Ok(DeclareTransaction::new(
+                sn_api_declare,
+                starknet_api::transaction::TransactionHash(txn_hash.into()),
+                ContractClass::Cairo1(self.contract_class.clone()).try_into()?,
+            )?)
+        }
```
Comment (FabijanC) [inline crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v2.rs:92]: This if-else logic is repeated in at least three places. Could it be extracted to a common function?

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v1.rs (hunk 1 above)
Comment (marioiordanov) [reply crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v1.rs:76]: I questioned myself for it, but checked out the transactions versions and each transaction starts at version 1

File: crates/starknet-devnet-core/src/starknet/add_invoke_transaction.rs [hunk 3]
```diff
@@ -169,6 +160,35 @@ mod tests {
...
+            BroadcastedInvokeTransaction::V3(ref mut v3) => {
+                v3.common.version = (FieldElement::from(3u8) + QUERY_VERSION_OFFSET).into();
+            }
+            _ => {
+                panic!("Wrong transaction type");
+            }
+        }
+
+        let txn_err = Starknet::default().add_invoke_transaction(invoke_transaction).unwrap_err();
+        println!("{:?}", txn_err);
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-devnet-core/src/starknet/add_invoke_transaction.rs]: [2e9b17d](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/414/commits/2e9b17d2e476d455d509e8e0010709e0e07f9f84)

File: crates/starknet-devnet-core/src/starknet/add_deploy_account_transaction.rs [hunk 4]
```diff
@@ -154,6 +160,29 @@ mod tests {
...
+        deploy_account_transaction.common.version =
+            (FieldElement::from(3u8) + QUERY_VERSION_OFFSET).into();
+
+        let txn_err = Starknet::default()
+            .add_deploy_account_transaction(BroadcastedDeployAccountTransaction::V3(
+                deploy_account_transaction,
+            ))
+            .unwrap_err();
+
+        println!("{:?}", txn_err);
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-devnet-core/src/starknet/add_deploy_account_transaction.rs]: [2e9b17d](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/414/commits/2e9b17d2e476d455d509e8e0010709e0e07f9f84)

File: crates/starknet-devnet-core/src/starknet/add_deploy_account_transaction.rs [hunk 5]
```diff
@@ -14,77 +12,84 @@ use super::Starknet;
...
+        }
+    };
 
-    if !starknet.state.is_contract_declared(broadcasted_deploy_account_transaction.class_hash) {
-        return Err(Error::StateError(crate::error::StateError::NoneClassHash(
-            broadcasted_deploy_account_transaction.class_hash,
-        )));
+    if blockifier_deploy_account_transaction.only_query {
+        return Err(Error::UnsupportedAction {
+            msg: "Only query transactions are not supported".to_string(),
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-devnet-core/src/starknet/add_deploy_account_transaction.rs]: [92b86c3](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/414/commits/92b86c39fdcf1ba9ea6c3a2f33ab0b956490ea77)

File: crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v2.rs (hunk 2 above)
Comment (marioiordanov) [reply crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v2.rs:92]: you mean the if condition -> self.common.version > QUERY_VERSION_OFFSET ?

Comment (FabijanC) [reply crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v2.rs:92]: The common function I had in mind could accept `(version, tx, hash, class)` do the if-else logic based on version, and return `Result<DeclareTransaction, ...>`. Would that be an improvement?

Comment (marioiordanov) [reply crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v2.rs:92]: On which struct this function will be implemented

Comment (marioiordanov) [reply crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v2.rs:92]: I thought of only extracting the comparison in a function to keep the logic in 1 place

Comment (FabijanC) [reply crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v2.rs:92]: yes, I had a regular function (non-method) in mind

Comment (marioiordanov) [reply crates/starknet-devnet-types/src/rpc/transactions/broadcasted_declare_transaction_v2.rs]: refactored

**Review by FabijanC — APPROVED** [review]
With only four lines added (+843 -839), the bug is gone. Looks good!

**Pull request merged**